nodeipam: align eTP=Cluster to kubernetes cloud-providers service lb (#31406)
cilium-team/cilium: Documentation/network/node-ipam.rst

is especially useful if you don't control the network you are running on and can't
use either the L2 or BGP capabilities of Cilium.

It works by getting the Node addresses of the selected Nodes and advertising them.
It will respect the ``.spec.ipFamilies`` to decide if IPv4 or IPv6 addresses
shall be used and will use the ``ExternalIP`` addresses if any or the
``InternalIP`` addresses otherwise.

> If the Service has ``.spec.externalTrafficPolicy`` sets to ``Cluster`` Node IPAM
> will consider all your nodes as candidate for selection. Or else if
> ``.spec.externalTrafficPolicy`` is set to ``Local``, Node IPAM will consider
> all the Pods selected by the Service (via their EndpointSlices) as candidate.

Reviewer: ```suggestion
If the Service has ``.spec.externalTrafficPolicy`` set to ``Cluster``, Node IPAM
considers all nodes as candidates for selection. Otherwise, if
``.spec.externalTrafficPolicy`` is set to ``Local``, then Node IPAM considers
all the Pods selected by the Service (via their EndpointSlices) as candidates.
```

.. warning::
    Node IPAM does work properly if ``.spec.externalTrafficPolicy`` is set

>     to ``Local`` but no EndpointSlice (or dummy EndpointSlice) are linked to

Reviewer: Missed the `are/is` on my previous review; and it looks like we lost the `not` when updating this paragraph, which is more annoying :slightly_smiling_face:.

```suggestion
    Node IPAM does not work properly if ``.spec.externalTrafficPolicy`` is set
    to ``Local`` but no EndpointSlice (or dummy EndpointSlice) is linked to
```

Author: Oups my bad indeed :sweat:

> .. warning::
>     Node IPAM will not work properly ``.spec.externalTrafficPolicy`` is set
>     to ``Local`` and that no EndpointSlice (or dummy EndpointSlice) are linked to
>     the corresponding Service.

Reviewer: ```suggestion
.. warning::
    Node IPAM does not work properly if ``.spec.externalTrafficPolicy`` is set
    to ``Local`` but no EndpointSlice (or dummy EndpointSlice) are linked to
    the corresponding Service.
```

>     As a result, you **cannot** use ``.spec.externalTrafficPolicy`` sets to ``Local``
>     with Cilium GatewayAPI or Cilium Ingress (although most GatewayAPI or Ingress
>     controller implementations should work/be compatible).

Reviewer: ```suggestion
    As a result, you **cannot** set ``.spec.externalTrafficPolicy`` to ``Local``
    with Cilium GatewayAPI or Cilium Ingress (although most GatewayAPI or Ingress
    controller implementations should work/be compatible).
```

`although most GatewayAPI or Ingress controller implementations should work/be compatible` -> Does this mean implementations other than Cilium? Why?

Author: Yep most implementation other than Cilium should work as Cilium implementation has a dummy endpoint (see this: https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-ingress-service.yaml#L58) so we cannot see the actual Endpoints of Cilium envoy in there. AFAIK most other ingress implementation are not doing that (not sure about GatewayAPI but I would assume it's similar-ish?).

This was the trigger for me to look into this btw as before this PR Node IPAM would always do that, so it was fully incompatible with Cilium ingress/GatewayAPI. Then I was kinda surprised this would work with other implementations of service LB and I looked into what is done with cloud providers etc.

Reviewer: Ok, thanks for the details! It would be nice to make the comment more precise in that case, by mentioning that “implementations other than Cilium should work”, and maybe summarising where the limitation stems from.

Author: I tried to be a bit more more explicit there, let me know if this seems correct to you!

> To restrict the Nodes that should listen for incoming traffic, you may add an
> annotation, ``io.cilium.nodeipam/match-node-labels``, to the Service. The value

Reviewer: ```suggestion
To restrict the Nodes that should listen for incoming traffic, add
annotation ``io.cilium.nodeipam/match-node-labels`` to the Service. The value
```

is especially useful if you don't control the network you are running on and can't
use either the L2 or BGP capabilities of Cilium.

It works by getting the Node addresses of the selected Nodes and advertising them.
It will respect the ``.spec.ipFamilies`` to decide if IPv4 or IPv6 addresses
shall be used and will use the ``ExternalIP`` addresses if any or the
``InternalIP`` addresses otherwise.

If the Service has ``.spec.externalTrafficPolicy`` set to ``Cluster``, Node IPAM
considers all nodes as candidates for selection. Otherwise, if
``.spec.externalTrafficPolicy`` is set to ``Local``, then Node IPAM considers
all the Pods selected by the Service (via their EndpointSlices) as candidates.

.. warning::
    Node IPAM does not work properly if ``.spec.externalTrafficPolicy`` is set
    to ``Local`` but no EndpointSlice (or dummy EndpointSlice) is linked to
    the corresponding Service.

    As a result, you **cannot** set ``.spec.externalTrafficPolicy`` to ``Local``
    with Cilium implementations of GatewayAPI/Ingress as Cilium currently uses
    a dummy Endpoints for the Service LoadBalancer (see

>     <https://github.com/cilium/cilium/blob/495f228ad8791c89f0851e0abbad90f09b136f80/install/kubernetes/cilium/templates/cilium-ingress-service.yaml#L58>`_).

Reviewer: ```suggestion
    As a result, you **cannot** set ``.spec.externalTrafficPolicy`` to ``Local``
    with the Cilium implementations for GatewayAPI or Ingress, because Cilium currently uses
    a dummy Endpoints for the Service LoadBalancer (see
    <https://github.com/cilium/cilium/blob/495f228ad8791c89f0851e0abbad90f09b136f80/install/kubernetes/cilium/templates/cilium-ingress-service.yaml#L58>`_).
```

>     Only the Cilium implementations are known to be broken in this specific configuration, expects
>     most other implementations to work and if it doesn't check if the matching EndpointSlices
>     looks correct and/or try setting ``.spec.externalTrafficPolicy`` to ``Cluster``.

Reviewer: ```suggestion
    Only the Cilium implementation is known to be affected by this limitation. Most other implementations are expected to work with this configuration. If they don't, check if the matching EndpointSlices
    look correct and/or try setting ``.spec.externalTrafficPolicy`` to ``Cluster``.
```
